### ISSUE TITLE
json: decoding of exponential notation

### ISSUE THREAD

**Issue by alfredh** [creytiv/re#69]
during a security audit it has been pointed out that the decoding of
JSON with a very large exponent can take a very long time, blocking the
process for several minutes.

example JSON strings:

```
{
  "test" : "0E999999999999999999999999"
}
```

**Comment by alfredh** [comment]
testing values and execution time:

```
0E99999999    0.2 seconds
0E999999999  2.3 seconds
0E9999999999 23 seconds
```

**Comment by alfredh** [comment]
the following solution was suggested by Chris Owen:
(in src/json/decode.c)

```
static inline double mypower10(uint64_t e)
{
	double n, p;

	n = 1.0;
	p = 10.0;
	while (e > 0) {
		if (e & 1) {
			n *= p;
		}
		p *= p;
		e >>= 1;
	}
	return n;
}
```

**Comment by alfredh** [comment]
I have tested with a couple of different solutions, and measured the performance.
The input JSON contains some numbers with exponent, and the decoding is
repeated 10000 times.

1. use `pow` from libm: 160049 usec
   (a disadvantage here is adding a dep to libm)

2. use patch from Chris: 160883 usec

3. in is_number use `strtol` instead: 305448 usec


The slowest solution is number 3 (use strol).

Solution number 1 and 2 has almost the same performance.

@richaas if you agree, I can prepare a PR for solution number 2.


here is the test JSON file:

```
{
  "array" : [
    0e9,
    0e99,
    0e999,
    0e9999,
    0e99999,
    0e999999,
    0e9999999,
    0e99999999,
    0e999999999,
    0e9999999999,

    1e9,
    1e99,
    1e999,
    1e9999,
    1e99999,
    1e999999,
    1e9999999,
    1e99999999,
    1e999999999,
    1e9999999999,

    2e10,
    2e100,
    2e1000,
    2e10000,
    2e100000,
    2e1000000,
    2e10000000,
    2e100000000,
    2e1000000000,
    2e10000000000,

  ]
}
```

**Comment by alfredh** [comment]
here is the PR: https://github.com/creytiv/re/pull/88

currently there is no checks in the code for boundaries/overflow.
both the base number and the exponent number is built up by parsing in reverse
the string containing [0-9].

**Comment by richaas** [comment]
Issue resolved in PR #88
